packages/beacon-node/src/sync/blockInputSync.ts

```diff
@@ -0,0 +1,617 @@
+        const rootHex = prettyBytes(isPendingBlockInput(block) ? block.blockInput.blockRootHex : block.rootHex);
+        this.logger.debug("Unexpected error - downloadBlock", {rootHex}, e);
+      });
+    }
+  };
+
+  private downloadBlock = async (block: BlockInputSyncCacheItem, connectedPeers: PeerIdStr[]): Promise<void> => {
+    if (block.status !== PendingBlockInputStatus.pending) {
+      return;
+    }
+    block.status = PendingBlockInputStatus.fetching;
+
+    this.logger.verbose("BlockInputSync.downloadBlock", getLogMeta(block, this.pendingBlocks));
+
+    const res = await wrapError(this.downloadBlockInputByRoot(block, connectedPeers));
+
+    if (res.err) {
+      this.metrics?.blockInputSync.downloadError.inc();
+      // Give up on this block and assume it does not exist, penalizing all peers as if it was a bad block
+      this.logger.debug(
+        `Ignoring block that cannot be correctly downloaded after ${MAX_ATTEMPTS_PER_BLOCK} failed attempts`,
+        getLogMeta(block, this.pendingBlocks),
+        res.err
+      );
+      this.removeAndDownScoreAllDescendants(block);
+      return;
+    }
+
+    const pending = res.result;
+    const {blockInput} = pending;
+    this.pendingBlocks.set(blockInput.blockRootHex, pending);
+    this.metrics?.blockInputSync.downloadSuccess.inc();
+
+    pending.status = PendingBlockInputStatus.downloaded;
+    pending.timeSyncedSec = Date.now() / 1000;
+
+    const blockSlot = blockInput.getBlock().message.slot;
+    this.metrics?.blockInputSync.timeToSyncSec.observe(pending.timeSyncedSec - pending.timeAddedSec);
+    const delaySec = pending.timeSyncedSec - (this.chain.genesisTime + blockSlot * this.config.SECONDS_PER_SLOT);
+    this.metrics?.blockInputSync.elapsedTimeTillReceived.observe(delaySec);
+
+    const parentRootHex = blockInput.parentRootHex;
+    const parentInForkChoice = this.chain.forkChoice.hasBlock(fromHex(parentRootHex));
+    const finalizedSlot = this.chain.forkChoice.getFinalizedBlock().slot;
+
+    if (parentInForkChoice) {
+      // Bingo! Process block. Add to pending blocks anyway for recycle the cache that prevents duplicate processing
+      this.processBlock(res.result).catch((e) => {
+        this.logger.debug("Unexpected error - process newly downloaded block", {}, e);
+      });
+    } else if (blockSlot <= finalizedSlot) {
+      // the common ancestor of the downloading chain and canonical chain should be at least the finalized slot and
+      // we should found it through forkchoice. If not, we should penalize all peers sending us this block chain
+      // 0 - 1 - ... - n - finalizedSlot
+      //                \
+      //                parent 1 - parent 2 - ... - unknownParent block
+      this.logger.debug("Downloaded block is before finalized slot", {
+        finalizedSlot,
+        blockSlot,
+        parentRootHex,
+        ...blockInput.getLogMeta(),
+      });
+      this.removeAndDownScoreAllDescendants(block);
```

Comment:
why would we be downscoring all the peers who gave us (probably) valid data?

Eg: Peer A sends attestation for an old orphaned block that we haven't seen (near the finality window)
We ask Peer B, C, D for the block and blobs for that block.
They send us the block and blobs and some seconds pass and finalization happens
We see that the block slot is too old, downscore Peer B, C, D.